Handle blocks that can not be loaded.

hanslovsky opened the pull request:
Currently, blocks that can not be loaded (`N5Reader.readBlock` returns `null`) defaults to leaving the corresponding `Img` unchanged, i.e. it will hold the default value for the appropriate primitive type. As discussed in #2 it can be desirable to set the corresponding `Img` to a different value, or throw an exception. This PR 
 - adds a `Consumer< Img >` member to the `N5CellLoader` that let's the user specify how to handle unavailable blocks. The default behavior remains unchanged.
 - adds appropriate convenience methods to `N5Utils`

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -323,16 +325,118 @@ private static void cropBlockDimensions(
+	 *
+	 * @param n5
+	 * @param dataset
+	 * @param ioExceptionHandler
```
name discrepancy

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5CellLoader.java:
```diff
@@ -158,4 +192,24 @@ public void load( final SingleCellArrayImg< T, ? > cell )
+	 *
+	 * @return {@link Consumer} that does not do anything.
+	 */
+	public static < T > Consumer< T > noOpConsumer()
```
Remove and replace `N5CellLoader.noOpConsumer()` calls by `t -> {}` which is a lot shorter and readable or re-use same instance, i.e.
```java
private Consumer<T> noOpConsumer = t ->{};
public static <T> Consumer<T> noOpConsumer() {
  return noOpConsumer;
}

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -323,16 +325,118 @@ private static void cropBlockDimensions(
+	}
+
+	/**
+	 * Open an N5 dataset as a disk-cached LazyCellImg.  Note that this
```
{@link LazyCellImg}

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -34,6 +35,7 @@
 import net.imglib2.cache.img.DiskCachedCellImgOptions;
 import net.imglib2.cache.img.LoadedCellCacheLoader;
 import net.imglib2.cache.ref.SoftRefLoaderCache;
+import net.imglib2.img.Img;
```
Where does that come from?

hanslovsky replied:
[Here](https://github.com/saalfeldlab/n5-imglib2/pull/3/files/aae06c3187ba808a542545a3e4e83356ae4fb062#diff-35426c6c65fe6d01148b0db841c6b851R433)

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -323,16 +325,118 @@ private static void cropBlockDimensions(
+	}
+
+	/**
+	 * Open an N5 dataset as a disk-cached LazyCellImg.  Note that this
```
{@link LazyCellImg}

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -414,19 +518,21 @@ private static void cropBlockDimensions(
 	 *
 	 * @param n5
 	 * @param dataset
+	 * @param ioExceptionHandler
```
parameter name

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -505,24 +611,26 @@ private static void cropBlockDimensions(
 
 	/**
 	 * Open an N5 dataset as a disk-cached LazyCellImg.  Note that this
```
{@link LazyCellImg}

axtimwalde commented on src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java:
```diff
@@ -505,24 +611,26 @@ private static void cropBlockDimensions(
 	 *
 	 * @param n5
 	 * @param dataset
+	 * @param ioExceptionHandler
```
parameter name

hanslovsky commented:
Addressed the comments in 3b99782

This PR should be squashable into master (I just recently noticed that github offers squashing through the web interface).

Is there an eclipse formatter scheme for N5? I am using the imglib2-clean-up formatter (as a save action) and it introduced a lot of irrelevant formatting changes, and figuring out what part of the diff is relevant is an unnecessary waste of time.

axtimwalde commented:
Good point.  n5-imglib2 used imglib2-clean-up, but it shouldn't.  I delayed reformatting until after the next meaningful merge.  Saalfeld lab projects should use this style (that I finally stored somewhere) :

https://github.com/saalfeldlab/n5/blob/master/doc/n5-eclipse-style.xml

It's Java conventions with fixed tabs and some reduced spacing.

With modern fonts, I find the ImgLib2 style too space consuming.

axtimwalde commented:
Thanks!

axtimwalde commented:
(of course I just had to update the style...)

hanslovsky commented:
Resolved with #3, this can be closed now.
